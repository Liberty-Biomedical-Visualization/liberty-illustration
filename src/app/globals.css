@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
