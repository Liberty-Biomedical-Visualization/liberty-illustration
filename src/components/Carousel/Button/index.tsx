import {
  faChevronLeft,
  faChevronRight,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import AccessiblyHiddenText from "@/components/AccessiblyHiddenText";
import resolveClassNames from "@/lib/resolveClassNames";

export default function CarouselButton(props: Readonly<CarouselButtonProps>) {
  const { disabled, onClick, type } = props;
  const className = resolveClassNames(baseClassName, props.className);
  const icon = type === "next" ? faChevronRight : faChevronLeft;
  const text = type === "next" ? "Next image" : "Previous image";

  return (
    <button className={className} disabled={disabled} onClick={onClick}>
      <FontAwesomeIcon icon={icon} />
      <AccessiblyHiddenText>{text}</AccessiblyHiddenText>
    </button>
  );
}

export interface CarouselButtonProps {
  className?: string;
  disabled?: boolean;
  onClick: () => void;
  type: "next" | "previous";
}

const baseClassName =
  "active:bg-accent-200 bg-accent-500 disabled:bg-accent-700 hover:bg-accent-300 py-2 sm:px-4 sm:py-12 text-background transition-colors";
