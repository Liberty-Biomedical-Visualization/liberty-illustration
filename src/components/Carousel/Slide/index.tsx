import Image from "next/image";

import resolveClassNames from "@/lib/resolveClassNames";

export default function CarouselSlide(props: Readonly<CarouselSlideProps>) {
  const { image } = props;

  const className = resolveClassNames(
    "flex items-center min-w-full px-2 w-full",
    props.className,
  );

  return (
    <div className={className}>
      <Image
        alt={image.description}
        className="max-h-full max-w-full mx-auto object-contain"
        height={image.height}
        src={image.src}
        title={image.title}
        width={image.width}
      />
    </div>
  );
}

export interface CarouselSlideProps {
  className?: string;
  image: ImageData;
}

interface ImageData {
  description: string;
  height: number;
  src: string;
  title: string;
  width: number;
}
