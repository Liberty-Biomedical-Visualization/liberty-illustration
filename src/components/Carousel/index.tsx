"use client";

import {
  type MutableRefObject,
  useCallback,
  useEffect,
  useRef,
  useState,
} from "react";

import * as array from "@/lib/array";
import resolveClassNames from "@/lib/resolveClassNames";

import CarouselButton from "./Button";
import CarouselStage, { type CarouselStageProps } from "./Stage";

export default function Carousel(props: Readonly<CarouselProps>) {
  const { images, slideDuration, transitionDuration } = props;

  const advanceSlideIntervalId = useRef<number | null>(null);
  const transitionTimeoutId = useRef<number | null>(null);

  const [currentIndex, setCurrentIndex] = useState(0);
  const [transition, setTransition] =
    useState<CarouselStageProps["transition"]>(null);

  const nextIndex = calculateIndex(currentIndex, +1, images.length);
  const previousIndex = calculateIndex(currentIndex, -1, images.length);

  const changeSlide = useCallback(
    (transition: CarouselStageProps["transition"], index: number) => {
      setTransition(transition);
      setTransitionTimeout(
        transitionDuration,
        index,
        setCurrentIndex,
        setTransition,
        transitionTimeoutId,
      );
    },
    [transitionDuration],
  );

  const advanceSlide = useCallback(
    () => changeSlide("next", nextIndex),
    [changeSlide, nextIndex],
  );

  const regressSlide = () => changeSlide("previous", previousIndex);

  useEffect(
    () =>
      initializeSlideshow(
        advanceSlide,
        advanceSlideIntervalId,
        images.length,
        slideDuration,
      ),
    [advanceSlide, images.length, slideDuration],
  );

  useEffect(() => {
    const cleanUpTransition = () => clearTransitionTimeout(transitionTimeoutId);
    return cleanUpTransition;
  }, []);

  if (images.length === 0) {
    return null;
  }

  const handleNextButtonClick = () =>
    handleButtonClick(
      advanceSlide,
      advanceSlideIntervalId,
      advanceSlide,
      slideDuration,
    );

  const handlePreviousButtonClick = () =>
    handleButtonClick(
      advanceSlide,
      advanceSlideIntervalId,
      regressSlide,
      slideDuration,
    );

  const className = resolveClassNames("flex justify-center", props.className);
  const buttonsAreDisplayed = images.length > 1;
  const buttonsAreDisabled = transition !== null;
  const currentImage = array.atChecked(images, currentIndex);
  const nextImage = array.atChecked(images, nextIndex);
  const previousImage = array.atChecked(images, previousIndex);

  return (
    <div className={className}>
      {buttonsAreDisplayed && (
        <CarouselButton
          className="my-auto"
          disabled={buttonsAreDisabled}
          onClick={handlePreviousButtonClick}
          type="previous"
        />
      )}
      <CarouselStage
        currentImage={currentImage}
        nextImage={nextImage}
        previousImage={previousImage}
        transition={transition}
        transitionDuration={transitionDuration}
      />
      {buttonsAreDisplayed && (
        <CarouselButton
          className="my-auto"
          disabled={buttonsAreDisabled}
          onClick={handleNextButtonClick}
          type="next"
        />
      )}
    </div>
  );
}

export interface CarouselProps {
  className?: string;
  images: readonly CarouselStageProps["currentImage"][];
  slideDuration: number | null;
  transitionDuration: CarouselStageProps["transitionDuration"];
}

function calculateIndex(
  currentIndex: number,
  delta: number,
  indexCount: number,
) {
  return (currentIndex + delta + indexCount) % indexCount;
}

function clearAdvanceSlideInterval(intervalId: TimerId) {
  if (intervalId.current !== null) {
    window.clearInterval(intervalId.current);
    intervalId.current = null;
  }
}

function clearTransitionTimeout(timeoutId: TimerId) {
  if (timeoutId.current !== null) {
    window.clearTimeout(timeoutId.current);
    timeoutId.current = null;
  }
}

function initializeSlideshow(
  advanceSlide: ChangeSlide,
  intervalId: TimerId,
  slideCount: number,
  slideDuration: number | null,
) {
  if (slideCount > 1 && slideDuration !== null) {
    setAdvanceSlideInterval(advanceSlide, intervalId, slideDuration);
    const finalizeSlideshow = () => clearAdvanceSlideInterval(intervalId);
    return finalizeSlideshow;
  }
}

function setAdvanceSlideInterval(
  advanceSlide: ChangeSlide,
  intervalId: TimerId,
  slideDuration: number,
) {
  intervalId.current = window.setInterval(advanceSlide, slideDuration);
}

function setTransitionTimeout(
  duration: number,
  index: number,
  setCurrentIndex: (index: number) => void,
  setTransition: (transition: null) => void,
  timeoutId: TimerId,
) {
  timeoutId.current = window.setTimeout(() => {
    setCurrentIndex(index);
    setTransition(null);
  }, duration);
}

function handleButtonClick(
  advanceSlide: ChangeSlide,
  advanceSlideIntervalId: TimerId,
  changeSlideOnClick: ChangeSlide,
  slideDuration: number | null,
) {
  clearAdvanceSlideInterval(advanceSlideIntervalId);
  changeSlideOnClick();

  if (slideDuration !== null) {
    setAdvanceSlideInterval(
      advanceSlide,
      advanceSlideIntervalId,
      slideDuration,
    );
  }
}

type ChangeSlide = () => void;
type TimerId = MutableRefObject<number | null>;
