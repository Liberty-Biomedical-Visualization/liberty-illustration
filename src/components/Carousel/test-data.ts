export const imageA = {
  description: "An imaginary image identified as A and used for testing.",
  height: 100,
  src: "https://foo.com/a.jpg",
  title: "Image A",
  width: 100,
} as const;

export const imageB = {
  description: "An imaginary image identified as B and used for testing.",
  height: 100,
  src: "https://foo.com/b.jpg",
  title: "Image B",
  width: 100,
} as const;

export const imageC = {
  description: "An imaginary image identified as C and used for testing.",
  height: 100,
  src: "https://foo.com/c.jpg",
  title: "Image C",
  width: 100,
} as const;
