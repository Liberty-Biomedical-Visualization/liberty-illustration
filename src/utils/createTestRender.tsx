import { render } from "@testing-library/react";
import type { JSX, JSXElementConstructor } from "react";

/**
 * Returns a function that allows rendering Component with partial props while
 * falling back to defaultProps for testing purposes.
 */
export default function createTestRender<C extends JSXElementConstructor<P>, P>(
  Component: JSXElementConstructor<P> & C,
  defaultProps: JSX.LibraryManagedAttributes<C, P>,
) {
  return function (props: Partial<P> = {}) {
    const resolvedProps: JSX.IntrinsicAttributes &
      JSX.LibraryManagedAttributes<C, P> = { ...defaultProps, ...props };
    return render(<Component {...resolvedProps} />);
  };
}
