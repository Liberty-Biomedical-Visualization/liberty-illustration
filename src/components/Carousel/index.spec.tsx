import { afterEach, describe, expect, it, jest } from "@jest/globals";
import { act, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import createTestRender from "@/utils/createTestRender";

import { imageA, imageB, imageC } from "./test-data";
import Carousel from ".";

describe("Carousel", () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  it("should disable the next button during a transition", async () => {
    jest.useFakeTimers();
    const slideDuration = 300;
    const transitionDuration = 100;
    renderCarousel({ slideDuration, transitionDuration });
    const button = screen.getByRole("button", { name: "Next image" });
    expect(button).not.toBeDisabled();

    await act(() => jest.advanceTimersByTimeAsync(slideDuration));
    expect(button).toBeDisabled();

    await act(() => jest.advanceTimersByTimeAsync(transitionDuration));
    expect(button).not.toBeDisabled();
  });

  it("should disable the previous button during a transition", async () => {
    jest.useFakeTimers();
    const slideDuration = 300;
    const transitionDuration = 100;
    renderCarousel({ slideDuration, transitionDuration });
    const button = screen.getByRole("button", { name: "Previous image" });
    expect(button).not.toBeDisabled();

    await act(() => jest.advanceTimersByTimeAsync(slideDuration));
    expect(button).toBeDisabled();

    await act(() => jest.advanceTimersByTimeAsync(transitionDuration));
    expect(button).not.toBeDisabled();
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderCarousel({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  describe("when passed zero images", () => {
    const images = [] as const;

    it("should not display any slides", () => {
      renderCarousel({ images });
      const anySlide = screen.queryByRole("img");
      expect(anySlide).toBeNull();
    });

    it("should not display the next button", () => {
      renderCarousel({ images });
      const button = screen.queryByRole("button", { name: "Next image" });
      expect(button).toBeNull();
    });

    it("should not display the previous button", () => {
      renderCarousel({ images });
      const button = screen.queryByRole("button", { name: "Previous image" });
      expect(button).toBeNull();
    });

    it("should match the snapshot", () => {
      const result = renderCarousel({ images });
      expect(result.container).toMatchSnapshot();
    });
  });

  describe("when passed one image", () => {
    const images = [imageA] as const;

    it("should display the current slide", () => {
      renderCarousel({ images });
      const currentSlide = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(currentSlide).toBeVisible();
    });

    it("should not advance slides when the slideDuration elapses", async () => {
      jest.useFakeTimers();
      const slideDuration = 300;
      renderCarousel({ images, slideDuration });
      const slideBeforeElapse = screen.getByRole("img", {
        name: imageA.description,
      });

      await act(() => jest.advanceTimersByTimeAsync(slideDuration));

      const slideAfterElapse = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeElapse).toBe(slideAfterElapse);
    });

    it("should not display the next button", () => {
      renderCarousel({ images });
      const button = screen.queryByRole("button", { name: "Next image" });
      expect(button).toBeNull();
    });

    it("should not display the previous button", () => {
      renderCarousel({ images });
      const button = screen.queryByRole("button", { name: "Previous image" });
      expect(button).toBeNull();
    });

    it("should match the snapshot", () => {
      const result = renderCarousel({ images });
      expect(result.container).toMatchSnapshot();
    });
  });

  describe("when passed two images", () => {
    const images = [imageA, imageB] as const;

    it("should display the current slide", () => {
      renderCarousel({ images });
      const currentSlide = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(currentSlide).toBeVisible();
    });

    it("should advance to the next slide when the slideDuration elapses", async () => {
      jest.useFakeTimers();
      const slideDuration = 300;
      renderCarousel({ images, slideDuration });
      const slideBeforeAdvance = screen.queryByRole("img", {
        name: imageB.description,
      });
      expect(slideBeforeAdvance).toBeNull();

      await act(() => jest.advanceTimersByTimeAsync(slideDuration));

      const slideAfterAdvance = screen.getByRole("img", {
        name: imageB.description,
      });
      expect(slideAfterAdvance).toBeVisible();
    });

    it("should hide the current slide when the slideDuration elapses and the transition is completed", async () => {
      jest.useFakeTimers();
      const slideDuration = 300;
      const transitionDuration = 100;
      renderCarousel({ images, slideDuration, transitionDuration });
      const slideBeforeTransition = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeTransition).toBeVisible();

      const waitDuration = slideDuration + transitionDuration;
      await act(() => jest.advanceTimersByTimeAsync(waitDuration));

      const slideAfterTransition = screen.queryByRole("img", {
        name: imageA.description,
      });
      expect(slideAfterTransition).toBeNull();
    });

    it("should advance to the next slide when the next button is clicked", async () => {
      renderCarousel({ images });
      const slideBeforeAdvance = screen.queryByRole("img", {
        name: imageB.description,
      });
      expect(slideBeforeAdvance).toBeNull();

      const user = userEvent.setup();
      const nextButton = screen.getByRole("button", { name: "Next image" });
      await user.click(nextButton);

      const slideAfterAdvance = screen.getByRole("img", {
        name: imageB.description,
      });
      expect(slideAfterAdvance).toBeVisible();
    });

    it("should hide the current slide when the next button is clicked and the transition is completed", async () => {
      jest.useFakeTimers();
      const transitionDuration = 100;
      renderCarousel({ images, transitionDuration });
      const slideBeforeTransition = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeTransition).toBeVisible();

      const user = userEvent.setup({
        advanceTimers: jest.advanceTimersByTimeAsync,
      });
      const nextButton = screen.getByRole("button", { name: "Next image" });
      await user.click(nextButton);
      await act(() => jest.advanceTimersByTimeAsync(transitionDuration));

      const slideAfterTransition = screen.queryByRole("img", {
        name: imageA.description,
      });
      expect(slideAfterTransition).toBeNull();
    });

    it("should regress to the previous slide when the previous button is clicked", async () => {
      renderCarousel({ images });
      const slideBeforeRegress = screen.queryByRole("img", {
        name: imageB.description,
      });
      expect(slideBeforeRegress).toBeNull();

      const user = userEvent.setup();
      const previousButton = screen.getByRole("button", {
        name: "Previous image",
      });
      await user.click(previousButton);

      const slideAfterRegress = screen.getByRole("img", {
        name: imageB.description,
      });
      expect(slideAfterRegress).toBeVisible();
    });

    it("should hide the current slide when the previous button is clicked and the transition is completed", async () => {
      jest.useFakeTimers();
      const transitionDuration = 100;
      renderCarousel({ images, transitionDuration });
      const slideBeforeTransition = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeTransition).toBeVisible();

      const user = userEvent.setup({
        advanceTimers: jest.advanceTimersByTimeAsync,
      });
      const previousButton = screen.getByRole("button", {
        name: "Previous image",
      });
      await user.click(previousButton);
      await act(() => jest.advanceTimersByTimeAsync(transitionDuration));

      const slideAfterTransition = screen.queryByRole("img", {
        name: imageA.description,
      });
      expect(slideAfterTransition).toBeNull();
    });

    it("should match the snapshot", () => {
      const result = renderCarousel({ images });
      expect(result.container).toMatchSnapshot();
    });
  });

  describe("when passed multiple images", () => {
    const images = [imageA, imageB, imageC] as const;

    it("should display the current slide", () => {
      renderCarousel({ images });
      const currentSlide = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(currentSlide).toBeVisible();
    });

    it("should advance to the next slide when the slideDuration elapses", async () => {
      jest.useFakeTimers();
      const slideDuration = 300;
      renderCarousel({ images, slideDuration });
      const slideBeforeAdvance = screen.queryByRole("img", {
        name: imageB.description,
      });
      expect(slideBeforeAdvance).toBeNull();

      await act(() => jest.advanceTimersByTimeAsync(slideDuration));

      const slideAfterAdvance = screen.getByRole("img", {
        name: imageB.description,
      });
      expect(slideAfterAdvance).toBeVisible();
    });

    it("should hide the current slide when the slideDuration elapses and the transition is completed", async () => {
      jest.useFakeTimers();
      const slideDuration = 300;
      const transitionDuration = 100;
      renderCarousel({ images, slideDuration, transitionDuration });
      const slideBeforeTransition = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeTransition).toBeVisible();

      const waitDuration = slideDuration + transitionDuration;
      await act(() => jest.advanceTimersByTimeAsync(waitDuration));

      const slideAfterTransition = screen.queryByRole("img", {
        name: imageA.description,
      });
      expect(slideAfterTransition).toBeNull();
    });

    it("should advance to the next slide when the next button is clicked", async () => {
      renderCarousel({ images });
      const slideBeforeAdvance = screen.queryByRole("img", {
        name: imageB.description,
      });
      expect(slideBeforeAdvance).toBeNull();

      const user = userEvent.setup();
      const nextButton = screen.getByRole("button", { name: "Next image" });
      await user.click(nextButton);

      const slideAfterAdvance = screen.getByRole("img", {
        name: imageB.description,
      });
      expect(slideAfterAdvance).toBeVisible();
    });

    it("should hide the current slide when the next button is clicked and the transition is completed", async () => {
      jest.useFakeTimers();
      const transitionDuration = 100;
      renderCarousel({ images, transitionDuration });
      const slideBeforeTransition = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeTransition).toBeVisible();

      const user = userEvent.setup({
        advanceTimers: jest.advanceTimersByTimeAsync,
      });
      const nextButton = screen.getByRole("button", { name: "Next image" });
      await user.click(nextButton);
      await act(() => jest.advanceTimersByTimeAsync(transitionDuration));

      const slideAfterTransition = screen.queryByRole("img", {
        name: imageA.description,
      });
      expect(slideAfterTransition).toBeNull();
    });

    it("should regress to the previous slide when the previous button is clicked", async () => {
      renderCarousel({ images });
      const slideBeforeRegress = screen.queryByRole("img", {
        name: imageC.description,
      });
      expect(slideBeforeRegress).toBeNull();

      const user = userEvent.setup();
      const previousButton = screen.getByRole("button", {
        name: "Previous image",
      });
      await user.click(previousButton);

      const slideAfterRegress = screen.getByRole("img", {
        name: imageC.description,
      });
      expect(slideAfterRegress).toBeVisible();
    });

    it("should hide the current slide when the previous button is clicked and the transition is completed", async () => {
      jest.useFakeTimers();
      const transitionDuration = 100;
      renderCarousel({ images, transitionDuration });
      const slideBeforeTransition = screen.getByRole("img", {
        name: imageA.description,
      });
      expect(slideBeforeTransition).toBeVisible();

      const user = userEvent.setup({
        advanceTimers: jest.advanceTimersByTimeAsync,
      });
      const previousButton = screen.getByRole("button", {
        name: "Previous image",
      });
      await user.click(previousButton);
      await act(() => jest.advanceTimersByTimeAsync(transitionDuration));

      const slideAfterTransition = screen.queryByRole("img", {
        name: imageA.description,
      });
      expect(slideAfterTransition).toBeNull();
    });

    it("should match the snapshot", () => {
      const result = renderCarousel({ images });
      expect(result.container).toMatchSnapshot();
    });
  });
});

const renderCarousel = createTestRender(Carousel, {
  images: [imageA, imageB, imageC],
  slideDuration: 5_000,
  transitionDuration: 1_000,
});
