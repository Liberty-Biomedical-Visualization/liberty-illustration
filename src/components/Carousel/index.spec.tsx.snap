// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel when passed multiple images should match the snapshot 1`] = `
<div>
  <div
    class="gap-4 grid grid-cols-2 grid-rows-[auto_1fr] sm:flex sm:gap-0 sm:justify-center"
  >
    <button
      class="active:bg-accent-200 bg-accent-500 col-span-1 disabled:bg-accent-700 hover:bg-accent-300 my-auto order-1 py-2 sm:order-none sm:px-4 sm:py-12 text-background transition-colors"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256 246.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
          fill="currentColor"
        />
      </svg>
      <span
        class="absolute border-0 h-px m-0 overflow-hidden p-0 w-px whitespace-nowrap"
      >
        Previous image
      </span>
    </button>
    <div
      class="col-span-2 flex flex-nowrap max-h-screen order-3 overflow-x-hidden sm:order-none sm:w-260"
    >
      <figure
        class="flex flex-col invisible items-center max-h-screen min-w-full opacity-0 px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as C and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fc.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fc.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fc.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image C"
          width="100"
        />
      </figure>
      <figure
        class="flex flex-col items-center max-h-screen min-w-full px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as A and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image A"
          width="100"
        />
      </figure>
      <figure
        class="flex flex-col invisible items-center max-h-screen min-w-full opacity-0 px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as B and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image B"
          width="100"
        />
      </figure>
    </div>
    <button
      class="active:bg-accent-200 bg-accent-500 col-span-1 disabled:bg-accent-700 hover:bg-accent-300 my-auto order-2 py-2 sm:order-none sm:px-4 sm:py-12 text-background transition-colors"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-right "
        data-icon="chevron-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M310.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L242.7 256 73.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
          fill="currentColor"
        />
      </svg>
      <span
        class="absolute border-0 h-px m-0 overflow-hidden p-0 w-px whitespace-nowrap"
      >
        Next image
      </span>
    </button>
  </div>
</div>
`;

exports[`Carousel when passed one image should match the snapshot 1`] = `
<div>
  <div
    class="gap-4 grid grid-cols-2 grid-rows-[auto_1fr] sm:flex sm:gap-0 sm:justify-center"
  >
    <div
      class="col-span-2 flex flex-nowrap max-h-screen order-3 overflow-x-hidden sm:order-none sm:w-260"
    >
      <figure
        class="flex flex-col invisible items-center max-h-screen min-w-full opacity-0 px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as A and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image A"
          width="100"
        />
      </figure>
      <figure
        class="flex flex-col items-center max-h-screen min-w-full px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as A and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image A"
          width="100"
        />
      </figure>
      <figure
        class="flex flex-col invisible items-center max-h-screen min-w-full opacity-0 px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as A and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image A"
          width="100"
        />
      </figure>
    </div>
  </div>
</div>
`;

exports[`Carousel when passed two images should match the snapshot 1`] = `
<div>
  <div
    class="gap-4 grid grid-cols-2 grid-rows-[auto_1fr] sm:flex sm:gap-0 sm:justify-center"
  >
    <button
      class="active:bg-accent-200 bg-accent-500 col-span-1 disabled:bg-accent-700 hover:bg-accent-300 my-auto order-1 py-2 sm:order-none sm:px-4 sm:py-12 text-background transition-colors"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256 246.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
          fill="currentColor"
        />
      </svg>
      <span
        class="absolute border-0 h-px m-0 overflow-hidden p-0 w-px whitespace-nowrap"
      >
        Previous image
      </span>
    </button>
    <div
      class="col-span-2 flex flex-nowrap max-h-screen order-3 overflow-x-hidden sm:order-none sm:w-260"
    >
      <figure
        class="flex flex-col invisible items-center max-h-screen min-w-full opacity-0 px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as B and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image B"
          width="100"
        />
      </figure>
      <figure
        class="flex flex-col items-center max-h-screen min-w-full px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as A and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fa.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image A"
          width="100"
        />
      </figure>
      <figure
        class="flex flex-col invisible items-center max-h-screen min-w-full opacity-0 px-2 relative right-full sm:justify-center w-full"
      >
        <img
          alt="An image identified as B and used for testing."
          class="max-h-full max-w-full mx-auto my-auto object-contain w-auto"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fb.jpg&w=256&q=75 2x"
          style="color: transparent;"
          title="Image B"
          width="100"
        />
      </figure>
    </div>
    <button
      class="active:bg-accent-200 bg-accent-500 col-span-1 disabled:bg-accent-700 hover:bg-accent-300 my-auto order-2 py-2 sm:order-none sm:px-4 sm:py-12 text-background transition-colors"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-right "
        data-icon="chevron-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M310.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L242.7 256 73.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
          fill="currentColor"
        />
      </svg>
      <span
        class="absolute border-0 h-px m-0 overflow-hidden p-0 w-px whitespace-nowrap"
      >
        Next image
      </span>
    </button>
  </div>
</div>
`;

exports[`Carousel when passed zero images should match the snapshot 1`] = `<div />`;
